Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp

```diff
@@ -0,0 +1,109 @@
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FEColorMatrixSkiaApplier.h"
+
+#if USE(SKIA)
+
+#include "FEColorMatrix.h"
+#include "FilterImage.h"
+#include "ImageBuffer.h"
+#include "NativeImage.h"
+#include <skia/core/SkCanvas.h>
+#include <skia/core/SkColorFilter.h>
+
+namespace WebCore {
+
+bool FEColorMatrixSkiaApplier::apply(const Filter&, const FilterImageVector& inputs, FilterImage& result) const
+{
+    ASSERT(inputs.size() == 1);
+    auto& input = inputs[0].get();
+
+    auto sourceImage = input.imageBuffer();
+    if (!sourceImage)
+        return false;
+
+    auto nativeImage = sourceImage->copyNativeImage();
+    if (!nativeImage || !nativeImage->platformImage())
+        return false;
+
+    auto values = FEColorMatrix::normalizedFloats(m_effect.values());
+    Vector<float> matrix;
+    float components[9];
+
+    switch (m_effect.type()) {
+    case ColorMatrixType::FECOLORMATRIX_TYPE_MATRIX:
+        matrix = values;
+        break;
+
+    case ColorMatrixType::FECOLORMATRIX_TYPE_SATURATE:
+        FEColorMatrix::calculateSaturateComponents(components, values[0]);
+        matrix = Vector<float>({
+            components[0], components[1], components[2], 0.0, 0.0,
+            components[3], components[4], components[5], 0.0, 0.0,
+            components[6], components[7], components[8], 0.0, 0.0,
+            0.0, 0.0, 0.0, 1.0, 0.0,
+        });
+        break;
+
+    case ColorMatrixType::FECOLORMATRIX_TYPE_HUEROTATE:
+        FEColorMatrix::calculateHueRotateComponents(components, values[0]);
+        matrix = Vector<float>({
+            components[0], components[1], components[2], 0.0, 0.0,
+            components[3], components[4], components[5], 0.0, 0.0,
+            components[6], components[7], components[8], 0.0, 0.0,
+            0.0, 0.0, 0.0, 1.0, 0.0,
+        });
+        break;
+
+    case ColorMatrixType::FECOLORMATRIX_TYPE_LUMINANCETOALPHA:
+        matrix = Vector<float>({
+            0.0,    0.0,    0.0,    0.0, 0.0,
+            0.0,    0.0,    0.0,    0.0, 0.0,
+            0.0,    0.0,    0.0,    0.0, 0.0,
+            0.2125, 0.7154, 0.0721, 0.0, 0.0,
+        });
+        break;
+
+    case ColorMatrixType::FECOLORMATRIX_TYPE_UNKNOWN:
+        return false;
+    }
+
+    auto* canvas = result.beginRecording();
+    {
+        SkPaint paint;
+        paint.setColorFilter(SkColorFilters::Matrix(matrix.data()));
+
+        canvas->drawImage(nativeImage->platformImage().get(), 0, 0, SkSamplingOptions(), &paint);
```

Comment:
There's a `SkCanvas::drawImage()` that receives a `sk_sp<SkImage>`, so you don't need the `get()`. You can also pass `{ }` instead of `SkSamplingOptions()`

Reply:
Done